src/uct/tcp/tcp_cm.c

```diff
@@ -21,69 +143,111 @@ unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
         goto err;
     }
 
-    iface->outstanding--;
+    if (uct_tcp_cm_send_conn_req(ep) != UCS_OK) {
+        ucs_debug("tcp_ep %p: unable to send connection request to %s",
+                  ep, ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr,
+                                       UCS_SOCKADDR_STRING_LEN));
+        return 0;
+    }
 
-    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTED);
+    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_WAIT_ACK);
+    uct_tcp_ep_mod_events(ep, EPOLLIN, EPOLLOUT);
 
     ucs_assertv((ep->tx.length == 0) && (ep->tx.offset == 0) &&
                 (ep->tx.buf == NULL), "ep=%p", ep);
-
-    return uct_tcp_ep_progress(ep, UCT_TCP_EP_CTX_TYPE_TX);
+    return 0;
 
 err:
-    iface->outstanding--;
     uct_tcp_ep_set_failed(ep);
     return 0;
 }
 
-void uct_tcp_cm_change_conn_state(uct_tcp_ep_t *ep, uct_tcp_ep_conn_state_t new_conn_state)
+unsigned uct_tcp_cm_conn_ack_rx_progress(uct_tcp_ep_t *ep)
 {
-    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface, uct_tcp_iface_t);
-    char str_local_addr[UCS_SOCKADDR_STRING_LEN], str_remote_addr[UCS_SOCKADDR_STRING_LEN];
-    uct_tcp_ep_conn_state_t old_conn_state;
+    uct_tcp_cm_conn_pkt_t conn_pkt;
 
-    old_conn_state = ep->conn_state;
-    ep->conn_state = new_conn_state;
+    if (uct_tcp_recv_blocking(ep->fd, &conn_pkt, sizeof(conn_pkt)) != UCS_OK) {
+        uct_tcp_ep_mod_events(ep, 0, EPOLLIN);
+        return 0;
+    }
 
-    if (!ucs_log_is_enabled(UCS_LOG_LEVEL_DEBUG)) {
-        return;
+    ucs_assertv(conn_pkt.event == UCT_TCP_CM_CONN_ACK, "ep=%p", ep);
+
+    uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTED);
+    uct_tcp_ep_mod_events(ep, EPOLLOUT, EPOLLIN);
+
+    ucs_assertv(ep->tx.buf == NULL, "ep=%p", ep);
+
+    /* Progress possibly pending TX operations */
+    return uct_tcp_ep_progress(ep, UCT_TCP_EP_CTX_TYPE_TX);
+}
+
+unsigned uct_tcp_cm_conn_req_rx_progress(uct_tcp_ep_t *ep)
+{
+    struct sockaddr peer_addr;
+    ucs_status_t status;
+
+    status = uct_tcp_cm_recv_conn_req(ep, &peer_addr);
+    if (status != UCS_OK) {
+        goto err;
     }
 
-    ucs_sockaddr_str((const struct sockaddr*)&iface->config.ifaddr,
-                     str_local_addr, UCS_SOCKADDR_STRING_LEN);
-    ucs_sockaddr_str(ep->peer_addr.addr, str_remote_addr, UCS_SOCKADDR_STRING_LEN);
+    uct_tcp_ep_addr_cleanup(&ep->peer_addr);
```

Comment:
no need for this call, `uct_tcp_ep_addr_init` should be enough

Reply:
why do we init `ep::peer_addr` before this func?

Reply:
add the comment into the code